flow/layers/container_layer.cc

```diff
@@ -10,6 +10,91 @@ namespace flutter {
 
 ContainerLayer::ContainerLayer() {}
 
+void ContainerLayer::AssignOldLayer(std::shared_ptr<ContainerLayer> old_layer) {
+  old_layer->old_layer_ = nullptr;
+  old_layer_ = old_layer;
+}
+
+void ContainerLayer::Diff(DiffContext* context, const Layer* old_layer) {
+  auto old_container = reinterpret_cast<const ContainerLayer*>(old_layer);
+  DiffContext::AutoSubtreeRestore subtree(context);
+  DiffChildren(context, old_container);
+  set_paint_region(context->CurrentSubtreeRegion());
+}
+
+void ContainerLayer::DiffChildren(DiffContext* context,
+                                  const ContainerLayer* old_layer) {
+  if (context->IsSubtreeDirty()) {
+    for (auto& layer : layers_) {
+      layer->Diff(context, nullptr);
+    }
+    return;
+  }
+  assert(old_layer);
+
+  const auto& prev_layers = old_layer->layers_;
+
+  // first mismatched element
+  int new_children_top = 0;
+  int old_children_top = 0;
+
+  // last mismatched element
+  int new_children_bottom = layers_.size() - 1;
+  int old_children_bottom = prev_layers.size() - 1;
+
+  while ((old_children_top <= old_children_bottom) &&
+         (new_children_top <= new_children_bottom)) {
+    if (!layers_[new_children_top]->CanDiff(
+            prev_layers[old_children_top].get())) {
+      break;
+    }
+    ++new_children_top;
+    ++old_children_top;
+  }
+
+  while ((old_children_top <= old_children_bottom) &&
+         (new_children_top <= new_children_bottom)) {
+    if (!layers_[new_children_top]->CanDiff(
+            prev_layers[old_children_top].get())) {
+      break;
+    }
+    --new_children_bottom;
+    --old_children_bottom;
+  }
+
+  // old layers that don't match
+  for (int i = old_children_top; i <= old_children_bottom; ++i) {
+    auto layer = prev_layers[i];
+    DiffContext::AutoSubtreeRestore subtree(context);
+    context->MarkSubtreeDirty(layer->paint_region());
```

Comment:
Would this be easier if AddDamage were public? It seems to me that doing an autorestore per removed child is more work than we really need here.